[5.4] Add @errors blade directive

m1guelpf opened the pull request:
This Pull Requests adds an @errors blade directive.

![](https://pbs.twimg.com/media/DEiqgoWUIAAUko5.jpg:large)
![](https://pbs.twimg.com/media/DEiqiYuUMAE23IR.jpg:large)
![](https://pbs.twimg.com/media/DEiqmmhUQAA9FCx.jpg:large)

Original idea by @paulredmond in Twitter.

## Structure

I feel like the oneliner of the implementation is very big. Would it be better to create a errors() helper and move the logic there?
ping @themsaid @taylorotwell

m1guelpf committed:
Add @errors blade directive

Original idea by @paulredmond in Twitter: https://twitter.com/paulredmond/status/885145291229249537/

browner12 commented:
i think this belongs in userland

m1guelpf commented:
@browner12 I think every CRUD application could benefit from this at least one time.

devcircus commented:
Of course, but there's no end to these helpers that could be beneficial and they're very simple to implement.

m1guelpf committed:
Syntax

m1guelpf commented:
@devcircus Isn't Laravel's job to make developer's lifes easier?

browner12 commented:
yah, just seems like a slippery slope here

m1guelpf committed:
cs

paulredmond commented:
I appreciate the sentiment, but I agree the original idea was in my own projects.

I'm also not sure how to get at the view variables and I'd cache this with static var or something so t doesn't hit the session store every time.

m1guelpf commented:
@paulredmond That's why I was suggesting to move the logic to a helper. Will do it tomorrow if it keeps open

paulredmond commented:
Every time something is added to core it must be maintained indefinitely is a non-breaking release. That list can add up quick, especially around small view helpers. I hope you don't take it personally as bad code or a bad PR 💖

m1guelpf commented:
@paulredmond I understand, but the amount of times I've had to write that manually makes me think it'd be awesome to have it on core... Let's wait for @taylorotwell's review...

taylorotwell commented:
Before we talk about if it can be included can you please clean-up that ternary expression. I can't even read the code tbh. Thanks.

vlakoff commented:
This directive doesn't sound clear to me, it sounds more like an iterable list than a conditional. I prefer the current syntax.

m1guelpf committed:
Add errors helper

m1guelpf committed:
Move logic to helper

fernandobandeira commented:
Another thing is that we can change the error variable name if we have more than one form on the same page, this won't cover this usecase...

m1guelpf committed:
cs

m1guelpf commented:
@fernandobandeira What do you mean? I don't understand...

m1guelpf commented:
@taylorotwell Cleaned up. If you approve this, I can add tests :smile:

fernandobandeira commented:
@m1guelpf 

```php
return redirect('register')
            ->withErrors($validator, 'login');

{{ $errors->login->first('email') }}
```

There's no way to filter the error bag using this new directive.

m1guelpf commented:
@fernandobandeira Will work in this tomorrow (if this is still open...)

deleugpn commented:
What about error bags that differs from `default`?

m1guelpf commented:
@deleugpn See @fernandobandeira's comment above and my answer :smile:

taylorotwell commented:
I'm not super opposed to a short-cut for this syntax as I agree it is used a lot in Laravel. There are some things that would need to be addressed (non-default bag, etc.) as others have mentioned. Will keep it open for now as I'm not totally opposed to the idea.

vlakoff commented:
Of course adding an helper `errors()` would be misleading and cluttering. It doesn't return errors but a boolean, it should be named `has_errors()`. Even better, it shouldn't be added. Laravel helpers are already occupying so many generic words.

m1guelpf commented:
@vlakoff It's better than having the logic in blade... I will rename it to has_errors, though. Do you have any other ideas about where the logic could be?

vlakoff commented:
I would suggest this:
```php
    protected function compileErrors($key = null)
    {
        if (is_null($key)) {
            return "<?php if(session('errors') && session('errors')->any()): ?>";
        }
    
        return "<?php if(session('errors') && session('errors')->has({$key})): ?>";
    }
```
<br>By the way, you could have false positives if `session('errors')` is not null, but an empty MessageBag.

vlakoff commented:
And a slightly optimized alternative:

```php
    protected function compileErrors($key = null)
    {
        if (is_null($key)) {
            return "<?php if(($errors = session('errors')) && $errors->any()): ?>";
        }

        return "<?php if(($errors = session('errors')) && $errors->has({$key})): ?>";
    }
```

vlakoff commented:
Please disregard the optimized alternative. We don't want to pollute the view by adding/altering variables.

m1guelpf commented:
@vlakoff The problem is that the logic will be more complex than that when I add custom non-default bags, and also Blade caches aggresively. I think I'll keep the helper, and we'll see what Taylor thinks.

m1guelpf committed:
Rename helper to has_errors

m1guelpf committed:
Support custom error bags

m1guelpf committed:
cs

m1guelpf committed:
Fix false positive

m1guelpf commented:
I've renamed the helper, added support for custom error bags & fixed a false positive. This is now ready for review. If @taylorotwell approves this, I'll add tests.

fernandobandeira commented on src/Illuminate/View/Compilers/Concerns/CompilesConditionals.php:
```diff
@@ -119,4 +119,25 @@ protected function compileEndAuth()
+     */
+    protected function compileErrors($key = null)
+    {
+        return "<?php if(errors({$key})): ?>";
```
This is still using the old ``` errors ``` helper, also we can't pass ``` $bag ``` I think you could change ``` $key ``` to accept an array too. This way:

```

@errors
Has errors
@enderrors

@errors('name')
Name field has errors
@enderrors

@errors(['bag' => 'login'])
Login form has errors
@enderrors

@errors(['bag' => 'login', 'key' => 'name'])
Name field of login form has errors
@enderrors

m1guelpf committed:
Fix call to helper

vlakoff commented on src/Illuminate/Foundation/helpers.php:
```diff
@@ -432,6 +432,27 @@ function encrypt($value)
+    function has_errors($key = null, $bag = null)
+    {
+        $errors = session('errors');
+        if (is_null($errors) && ! session('errors')->any()) {
```
Logic is wrong. It should be `if (is_null($errors) || ! $errors->any())`

m1guelpf committed:
Merge branch '5.4' into errors-bla

m1guelpf committed:
Fix logic

vlakoff commented on src/Illuminate/Foundation/helpers.php:
```diff
@@ -432,6 +432,27 @@ function encrypt($value)
+    function has_errors($key = null, $bag = null)
+    {
+        $errors = session('errors');
+        if (if (is_null($errors) || ! $errors->any())) {
```
Please put the bottle down

m1guelpf committed:
typo

m1guelpf committed:
Order helper

vlakoff commented on src/Illuminate/Foundation/helpers.php:
```diff
@@ -443,7 +443,7 @@ function encrypt($value)
     {
         $errors = session('errors');
-        if (if (is_null($errors) || ! $errors->any())) {
+        if (is_null($errors) || ! $errors->any()) {
```
Also it always checks the default bag, even if a bag has been specified.

Example of fix:
`is_null($errors) || ! $errors->getBag($bag ?: 'default')->any()`

vlakoff replied:
Even better, at top of code: `$bag = $bag ?: 'default'`

Then for the rest of the code you can do: `$errors->$bag->...`

m1guelpf committed:
Use custom bags in first check

taylorotwell commented:
Please provide code examples of using all features, including custom bags, etc.

m1guelpf commented:
@taylorotwell There you have!
![errorhelper](https://user-images.githubusercontent.com/23558090/28320936-33fe1262-6bd2-11e7-987a-9ac93d86791c.png)

m1guelpf committed:
Fix tests

fernandobandeira commented on src/Illuminate/View/Compilers/Concerns/CompilesConditionals.php:
```diff
@@ -140,4 +140,25 @@ protected function compileEndIsset()
+     */
+    protected function compileErrors($expression = null)
+    {
+        return "<?php if(errors{$expression}): ?>";
```
Shouldn't it be ``` has_errors ``` ?
Also the docblock isn't matching the parameter name anymore
```
@param  string|null  $key
```

deleugpn commented:
@m1guelpf could you please try 

`@includeWhen(@errors('field'), 'my-error-field-component')`

m1guelpf committed:
Update component

deleugpn commented on src/Illuminate/Foundation/helpers.php:
```diff
@@ -470,6 +470,27 @@ function factory()
+     * @param  string|null  $bag
+     * @return bool
+     */
+    function has_errors($key = null, $bag = null)
```
`$bag = 'default'`.

`$errors->getBag($bag)->any()`

deleugpn replied:
`Null` is inconsistent with other parts of the framework. I think it should stay as `default` https://github.com/laravel/framework/blob/5.4/src/Illuminate/Support/ViewErrorBag.php#L23

deleugpn commented on src/Illuminate/Foundation/helpers.php:
```diff
@@ -470,6 +470,27 @@ function factory()
+        if (is_null($errors) || ! $errors->getBag($bag ?: 'default')->any()) {
+            return false;
+        } elseif ($key !== null) {
+            return is_null($bag) ? $errors->has($key) : $errors->$bag->has($key);
```
`$errors->getBag($bag)->has($key);`

deleugpn commented on src/Illuminate/Foundation/helpers.php:
```diff
@@ -470,6 +470,27 @@ function factory()
+    function has_errors($key = null, $bag = null)
+    {
+        $errors = session('errors');
+        if (is_null($errors) || ! $errors->getBag($bag ?: 'default')->any()) {
```
`$errors->getBag($bag)->any()`

m1guelpf committed:
Set default error bag by default

fernandobandeira reviewed (approved):
Seems way better 😄

vlakoff commented on src/Illuminate/Foundation/helpers.php:
```diff
@@ -470,6 +470,27 @@ function factory()
+     * @param  string|null  $bag
+     * @return bool
+     */
+    function has_errors($key = null, $bag = 'default')
```
Should be changed back to `$bag = null`, then add something like `$bag = $bag ?: 'default'` at beginning of code. One should be able to programmatically use the default bag by passing `null`.

deleugpn replied:
I disagree with this statement, but even if I agreed to it, it would have to be a separate PR to also change https://github.com/laravel/framework/blob/5.4/src/Illuminate/Support/ViewErrorBag.php#L23 for consistency.

vlakoff replied:
Thanks for pointing out this, indeed it should be consistent.

It's pretty standard to pass `null` for using the default value, so one doesn't have to remember if it's "default", "\_default", "\_\_default\_\_" or whatever. But indeed this is a different changeset.

m1guelpf replied:
@vlakoff I think I'll revert back to 'default' for consistency. You can submit a PR changing both if this gets merged.

m1guelpf replied:
@vlakoff Anyway, if you don't specify the second value, it'll assume 'default'. You don't need to remember it :smile:

vlakoff replied:
I said "programmatically", i.e. there are some cases you are forcibly specifying an argument.

vlakoff commented on src/Illuminate/Foundation/helpers.php:
```diff
@@ -470,6 +470,27 @@ function factory()
+        if (is_null($errors) || ! $errors->getBag($bag)->any()) {
+            return false;
+        } elseif ($key !== null) {
+            return $errors->$bag->has($key);
```
Pick `->getBag($bag)` or `->$bag`, but not both.

m1guelpf committed:
Add tests

m1guelpf commented:
@taylorotwell Everything is ready for review now :octocat:

taylorotwell commented:
I don't like being forced to pass null in order to check non-default bag. Feels clunky.

alexbowers commented:
Could it not be done as part of dot syntax?

`@errors('<bag>.array.syntax')` 

Checking against the array syntax first for default array dot syntax, and then falling back to using the first option as the name of a bag, if no value is found.

deleugpn commented:
Can people use `.` as the name of an input? If that doesn't get translated to array, then @alexbowers would not be feasible.
